### ISSUE TITLE
【训练技巧】功守道：NLP中的对抗训练 + PyTorch实现 | 瓦特兰蒂斯

### ISSUE THREAD

**Issue by fyubang** [fyubang/gitalk#25]
http://fyubang.com/2019/10/15/adversarial-train/ 

 本文分享一个“万物皆可盘”的NLP对抗训练实现，只需要四行代码即可调用。你值得拥有。  最近，微软的FreeLB-Roberta [1] 靠着对抗训练 (Adversarial Training) 在GLUE榜上超越了Facebook原生的Roberta，追一科技也用到了这个方法仅凭单模型 [2] 就在CoQA榜单中超过了人类，似乎“对抗训练”一下子变成了NLP任务的一把利器。刚好笔者最近也在看

**Comment by mujizi** [comment]
你好：
我看过了你很多文章，对我很有帮助。
请问tensorflow即插即用可以实现吗，我想试试看自己实现，还是有已经实现的仓库？

**Comment by fyubang** [comment]
> @mujizi
> 你好：
> 我看过了你很多文章，对我很有帮助。
> 请问tensorflow即插即用可以实现吗，我想试试看自己实现，还是有已经实现的仓库？

谢谢支持~
Tensorflow可以参考一下原作的实现：https://github.com/tensorflow/models/blob/e97e22dfcde0805379ffa25526a53835f887a860/research/adversarial_text/adversarial_losses.py

**Comment by mujizi** [comment]
作者你好：
在tensorflow中实现，我发现在embedding的restore这一步，tensorflow中的tensor是一个只读的对象，不能直接update原来的embedding值进行更新，请问有什么好办法吗，原作是adv_loss 和 loss直接分开算的不存在这个问题

**Comment by Billy-Zhoup** [comment]
您好：
想请问下，我可以比如说同时Virtual Adversarial Training 和 Adversarial Training么？
之前看一篇论文，说他同时用了这两种方法，但没有放代码实现。
那就是在正常grad基础上，把两种对抗训练的梯度都累加上去，然后再optimizer.step()？
谢谢您，麻烦了：）

**Comment by fyubang** [comment]
> @Billy-Zhoup
> 您好：
> 想请问下，我可以比如说同时Virtual Adversarial Training 和 Adversarial Training么？
> 之前看一篇论文，说他同时用了这两种方法，但没有放代码实现。
> 那就是在正常grad基础上，把两种对抗训练的梯度都累加上去，然后再optimizer.step()？
> 谢谢您，麻烦了：）

可以的。对的，累加grad就可以了

**Comment by guozhiyu** [comment]
我把Fast Gradient Method写成了tensorflow2.x版本的想应用在一个QA任务上，由于数据集比较大，我需要使用TPU， 但运行后发现这在tf2.0做并行训练的情况下很难实现，每个核都要给model改参数和还原参数，而模型的参数在多核间是共享的，这会让参数的改变顺序变得杂乱无章。

**Comment by euminds** [comment]
请问，这篇博客有提供完整的代码，还有数据集是什么？目前刚入坑NLP的对抗训练。另外，目前想用pytorch复现HotFlip: White-Box Adversarial Examples for Text Classiﬁcation的内容，实现对抗样本生成。

**Comment by fyubang** [comment]
> @euminds
> 请问，这篇博客有提供完整的代码，还有数据集是什么？目前刚入坑NLP的对抗训练。另外，目前想用pytorch复现HotFlip: White-Box Adversarial Examples for Text Classiﬁcation的内容，实现对抗样本生成。
> 

文章里的源码很完整了，数据集也写了，都是GLUE的数据集。

**Comment by zhengyanzhao1997** [comment]
> @mujizi
> 作者你好：
> 在tensorflow中实现，我发现在embedding的restore这一步，tensorflow中的tensor是一个只读的对象，不能直接update原来的embedding值进行更新，请问有什么好办法吗，原作是adv_loss 和 loss直接分开算的不存在这个问题

你好,请问您最后解决了吗～

**Comment by wmathor** [comment]
作者您好，首先感谢您的文章分享，写的非常好
其次我有个小疑问，第二个代码片中，采用了累加梯度，这个我可以理解，但还是感觉有点不太舒服，因为毕竟很少用到梯度累加。所以我思考了下，如果在第8行和第9行之间加一个`model.zero_grad()`或许会更好？这样的话就没有梯度累加了，但是我不太确定这样是否可行（代码应该不会报错，我主要是怀疑这样还有没有效果），因此希望见多识广的作者给点意见，谢谢啦

**Comment by DingLuBJUT** [comment]
感觉第二个代码中存在bug，最后跳出range(k)的时候进行了restore_grad，那么之后的参数更新就没有使用到累计的梯度信息

**Comment by fyubang** [comment]
> @DingLuBJUT
> 感觉第二个代码中存在bug，最后跳出range(k)的时候进行了restore_grad，那么之后的参数更新就没有使用到累计的梯度信息


> @DingLuBJUT
> 感觉第二个代码中存在bug，最后跳出range(k)的时候进行了restore_grad，那么之后的参数更新就没有使用到累计的梯度信息

您好，你可以再看看pgd的算法具体的流程，前面累计grad不是为了最终修改模型参数，只是为了给embedding加扰动。

**Comment by Coding-Zuo** [comment]
你好讲的很棒，很感谢。有提升！

**Comment by Chandler-Bing** [comment]
fgm.attack 函数中，para.grad为None，请问这个怎么解决的
